README
## This project is in need of a steward. If you are interested in helping maintain stjerm, please reach out.

Requirements
------------
pkgconfig, gtk2, vte

Ubuntu Requirements
-------------------
If you're compiling from source on Ubuntu, you'll need:

    sudo apt-get install build-essential automake libglib2.0-dev libvte-dev

Installation
------------
    ./autogen.sh
    ./configure
    make
    sudo make install


Running
----------
Simply run:
    stjerm

and all available options will be listed.


debian/changelog
stjerm (0.18-git1) trusty; urgency=medium

  * git rev. 3340620
  * added hideOnChildExit option
  * added ignoreDeleteEvent option
  * added infinite buffer for lines option (use negative numbers)
  * fix: read options cycle

 -- Domenico Martella (dom0) <domenico.martella@alcacoop.it>  Thu, 18 Dec 2014 17:38:42 +0100

stjerm (0.16-0ubuntu2) raring; urgency=low

  * Added missing libs to link, fixing FTBFS.

 -- Daniel T Chen <crimsun@ubuntu.com>  Mon, 15 Apr 2013 21:07:11 -0400

stjerm (0.16-0ubuntu1) raring; urgency=low

  * New upstream release.
  * Drop debian/patches/man-hyphen.patch (fixed upstream).
  * Bump Standards-Version to 3.9.3.

 -- Logan Rosen <logatronico@gmail.com>  Sat, 03 Nov 2012 16:30:44 -0400

stjerm (0.15-0ubuntu1) precise; urgency=low

  * New upstream release (LP: #553958).
  * Use debhelper 7 rules file.
  * Use dh-autoreconf to run autogen.sh.
  * Add build-dependencies: libglib2.0-dev, pkg-config.
  * Update homepage and watch file.
  * man-hyphen.patch: properly escape hyphens in manual page.
  * Bump debhelper/compat to 7.
  * Bump Standards-Version to 3.9.2.

 -- Ilya Barygin <randomaction@ubuntu.com>  Sat, 15 Oct 2011 21:47:28 +0400

stjerm (0.11-0ubuntu1) intrepid; urgency=low

  * Initial release (LP: #216603)

 -- Carol Meertens <mouz@xs4all.nl>  Tue, 06 Jul 2008 22:48:49 +0200



src/stjerm.h
/*
 * stjerm.h
 * This file is part of Stjerm
 *
 * Copyright (C) 2007-2010 - Kristopher Wilson, Stjepan Glavina and Markus Groß
 * 
 * Stjerm is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Stjerm is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Stjerm; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */

#ifndef __STJERM_H__
#define __STJERM_H__

#include <gtk/gtk.h>
#include <vte/vte.h>
#include <X11/Xlib.h>

#define TERM_ICONIFY_WINDOW    0
#define TERM_DEICONIFY_WINDOW  1
#define TERM_RAISE_WINDOW      2
#define TERM_LOWER_WINDOW      3
#define TERM_MAXIMIZE_WINDOW   4
#define TERM_RESTORE_WINDOW    5
#define TERM_REFRESH_WINDOW    6
#define TERM_RESIZE_WINDOW     7
#define TERM_MOVE_WINDOW       8

#define BORDER_THIN  0
#define BORDER_THICK 1
#define BORDER_NONE  2

#define POS_TOP         0
#define POS_BOTTOM      1
#define POS_LEFT        2
#define POS_RIGHT       3
#define POS_TOPLEFT     4
#define POS_TOPRIGHT    5
#define POS_BOTTOMLEFT  6
#define POS_BOTTOMRIGHT 7

#define TABS_NEVER  0
#define TABS_ONE    1
#define TABS_ALWAYS 2

#define OPTION_COUNT 32


// The following defines borrowed from gnome-terminal:

#define USERCHARS "-[:alnum:]"
#define USERCHARS_CLASS "[" USERCHARS "]"
#define PASSCHARS_CLASS "[-[:alnum:]\\Q,?;.:/!%$^*&~\"#'\\E]"
#define HOSTCHARS_CLASS "[-[:alnum:]]"
#define HOST HOSTCHARS_CLASS "+(\\." HOSTCHARS_CLASS "+)*"
#define PORT "(?:\\:[[:digit:]]{1,5})?"
#define PATHCHARS_CLASS "[-[:alnum:]\\Q_$.+!*,;@&=?/~#%\\E]"
#define PATHTERM_CLASS "[^\\Q]'.}>) \t\r\n,\"\\E]"
#define SCHEME "(?:news:|telnet:|nntp:|file:\\/|https?:|ftps?:|sftp:|webcal:)"
#define USERPASS USERCHARS_CLASS "+(?:" PASSCHARS_CLASS "+)?"
#define URLPATH   "(?:(/"PATHCHARS_CLASS"+(?:[(]"PATHCHARS_CLASS"*[)])*"PATHCHARS_CLASS"*)*"PATHTERM_CLASS")?"


typedef struct {
    char long_name[18];
    char short_name[12];
    char var_type[12];
    char desc[150];
} Option;

typedef enum {
    URL_TYPE_HTTP,
    URL_TYPE_EMAIL,
    URL_TYPE_NEWS
} UrlTypes;

typedef struct {
    const char *pattern;
    UrlTypes url_type;
    GRegexCompileFlags flags;
} RegExPattern;

typedef struct {
    int tag;
    char *text;
} CursorMatch;


// The following patterns borrowed from gnome-terminal:

static const RegExPattern uri_patterns[] = {
    {"news:[[:alnum:]\\Q^_{|}~!\"#$%&'()*+,./;:=?`\\E]+", URL_TYPE_NEWS, G_REGEX_CASELESS },
    {SCHEME "//(?:" USERPASS "\\@)?" HOST PORT URLPATH, URL_TYPE_HTTP, G_REGEX_CASELESS},
    {"(?:www|ftp)" HOSTCHARS_CLASS "*\\." HOST PORT URLPATH , URL_TYPE_HTTP, G_REGEX_CASELESS},
    {"(?:mailto:)?" USERCHARS_CLASS "[" USERCHARS ".]*\\@" HOSTCHARS_CLASS "+\\." HOST, URL_TYPE_EMAIL, G_REGEX_CASELESS},
};


extern void print_help(void);
extern void print_info(void);
extern void print_version(void);

extern void build_mainwindow(void);
extern GtkWidget* build_term(void);
extern void build_popupmenu(void);

extern void mainwindow_toggle(int sig);
extern void mainwindow_create_tab(void);
extern void mainwindow_close_tab(GtkWidget *);
extern void mainwindow_toggle_fullscreen(void);
extern void init_key(void);
extern void grab_key(void);
extern void wait_key(void);

extern void buildprefswindow(void);

extern void conf_init(void);
extern char* conf_get_font(void);
extern float conf_get_opacity(void);
extern GdkColor conf_get_bg(void);
extern GdkColor conf_get_fg(void);
extern int conf_get_border(void);
extern unsigned int conf_get_mod(void);
extern KeySym conf_get_key(void);
extern int conf_get_width(void);
extern int conf_get_height(void);
extern void conf_get_position(int*, int*);
extern int conf_get_scrollbar(void);
extern char* conf_get_shell(void);
extern char* conf_get_emulation(void);
extern int conf_get_lines(void);
extern int conf_get_show_tab(void);
extern char* conf_get_term_name(void);
extern GtkPositionType conf_get_tab_pos(void);
extern GdkColor* conf_get_color_palette(void);
extern gboolean conf_get_tab_fill(void);
extern gboolean conf_get_allow_bold(void);
extern GdkModifierType conf_get_key_mod(void);
extern gboolean conf_get_auto_hide(void);
extern char* conf_get_bg_image(void);
extern gboolean conf_get_allow_reorder(void);
extern gboolean conf_get_scroll_on_output();
extern gboolean conf_get_cursor_blink(void);
extern GdkColor conf_get_cursor_color(void);
extern VteTerminalCursorShape conf_get_cursor_shape(void);
extern gboolean conf_get_toggled(void);


#endif /* __STJERM_H__ */



src/mainwindow.c
/*
 * mainwindow.c
 * This file is part of Stjerm
 *
 * Copyright (C) 2007-2010 - Kristopher Wilson, Stjepan Glavina and Markus Groß
 * 
 * Stjerm is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Stjerm is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Stjerm; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */

#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <gtk/gtk.h>
#include <gdk/gdkx.h>
#include <gdk/gdkkeysyms.h>
#include <vte/vte.h>
#include <stdlib.h>
#include "stjerm.h"

extern GtkWidget *popupmenu;
extern GtkWidget *close_tab;
extern gboolean popupmenu_shown;

GtkWidget *mainwindow;
int activetab;
int tabcount;
GtkNotebook* tabbar;

Window mw_xwin;
static Display *dpy = 0;
Atom opacityatom;
gboolean screen_is_composited;
gboolean fullscreen;
gboolean toggled;


void build_mainwindow(void);
void mainwindow_toggle(int sig);
void mainwindow_create_tab(void);
void mainwindow_close_tab(GtkWidget *term);
void mainwindow_toggle_fullscreen(void);
int handle_x_error(Display *dpy, XErrorEvent *evt);

GtkWidget* mainwindow_get_terminal_at(int);

static GRegex **uri_regex;
static guint uri_regex_count;

static void mainwindow_reset_position(void);
static void mainwindow_focus_terminal(void);
static void mainwindow_show(GtkWidget*, gpointer);
static void mainwindow_focus_out_event(GtkWindow*, GdkEventFocus*, gpointer);
static gboolean mainwindow_expose_event(GtkWidget*, GdkEventExpose*, gpointer);
static void mainwindow_destroy(GtkWidget*, gpointer);
static void mainwindow_window_title_changed(VteTerminal *vteterminal,
    gpointer user_data);
static void mainwindow_switch_tab(GtkNotebook *notebook, GtkNotebookPage *page,
    guint page_num, gpointer user_data);
static void mainwindow_goto_tab(gint i);
static void mainwindow_next_tab(GtkWidget *widget, gpointer user_data);
static void mainwindow_prev_tab(GtkWidget *widget, gpointer user_data);
static void mainwindow_new_tab(GtkWidget *widget, gpointer user_data);
static void mainwindow_delete_tab(GtkWidget *widget, gpointer user_data);
static gint mainwindow_copy(GtkWidget *widget, gpointer user_data);
static gint mainwindow_paste(GtkWidget *widget, gpointer user_data);
static gint delete_event(GtkWidget *widget, GdkEvent  *event, gpointer data);

static gint mainwindow_tab_at_xy(GtkNotebook *notebook, gint abs_x, gint abs_y);
static void mainwindow_notebook_clicked(GtkWidget *widget, GdkEventButton *event, gpointer func_data);

void build_mainwindow(void)
{
    guint ic;
  
    uri_regex_count = G_N_ELEMENTS(uri_patterns);
    uri_regex = g_new0(GRegex*, uri_regex_count);

    for(ic = 0; ic < uri_regex_count; ++ic)
    {
        GError *error = NULL;
        
        uri_regex[ic] = g_regex_new(uri_patterns[ic].pattern,
            uri_patterns[ic].flags | G_REGEX_OPTIMIZE, 0, &error);
        
        if(error)
        {
            g_message("%s", error->message);
            g_error_free (error);
        }
    }

    mainwindow = gtk_window_new(GTK_WINDOW_TOPLEVEL);

    gtk_widget_set_app_paintable(mainwindow, TRUE);
    gtk_widget_set_size_request(mainwindow, conf_get_width(), conf_get_height());
    gtk_window_set_decorated(GTK_WINDOW(mainwindow), FALSE);
    gtk_window_set_skip_taskbar_hint(GTK_WINDOW(mainwindow), TRUE);
    gtk_window_set_skip_pager_hint(GTK_WINDOW(mainwindow), TRUE);
    gtk_window_set_resizable(GTK_WINDOW(mainwindow), TRUE);
    mainwindow_reset_position();

    fullscreen = FALSE;
    toggled = FALSE;

    GtkAccelGroup* accel_group;
    GClosure *new_tab, *delete_tab, *next_tab, *prev_tab, *delete_all,
        *maximize, *copy, *paste;

    GClosure *goto_tab_closure[10];
    long i;

    accel_group = gtk_accel_group_new();
    gtk_window_add_accel_group(GTK_WINDOW(mainwindow), accel_group);

    maximize = g_cclosure_new_swap(G_CALLBACK(mainwindow_toggle_fullscreen),
        NULL, NULL);
    gtk_accel_group_connect(accel_group, GDK_F11, 0,
        GTK_ACCEL_VISIBLE, maximize);

    new_tab = g_cclosure_new_swap(G_CALLBACK(mainwindow_new_tab), 
        NULL, NULL);
    gtk_accel_group_connect(accel_group, 't', conf_get_key_mod(),
        GTK_ACCEL_VISIBLE, new_tab);

    delete_tab = g_cclosure_new_swap(G_CALLBACK(mainwindow_delete_tab), 
        NULL, NULL);
    gtk_accel_group_connect(accel_group, 'w', conf_get_key_mod(),
        GTK_ACCEL_VISIBLE, delete_tab);

    next_tab = g_cclosure_new_swap(G_CALLBACK(mainwindow_next_tab), 
        NULL, NULL);
    gtk_accel_group_connect(accel_group, GDK_Page_Up, conf_get_key_mod(),
        GTK_ACCEL_VISIBLE, next_tab);

    prev_tab = g_cclosure_new_swap(G_CALLBACK(mainwindow_prev_tab), 
        NULL, NULL);
    gtk_accel_group_connect(accel_group, GDK_Page_Down, conf_get_key_mod(),
        GTK_ACCEL_VISIBLE, prev_tab);

    delete_all = g_cclosure_new_swap(G_CALLBACK(mainwindow_destroy), 
        NULL, NULL);
    gtk_accel_group_connect(accel_group, 'q', conf_get_key_mod(),
        GTK_ACCEL_VISIBLE, delete_all);

    /* tab hotkeys, inspired by Tilda -- thanks to castorinop for the patch */
    for(i = 0; i < 10; i++)
    {
        goto_tab_closure[i] = g_cclosure_new_swap(G_CALLBACK(mainwindow_goto_tab),
            (gpointer) i, NULL);
        gtk_accel_group_connect(accel_group, '0' + ((i+1)%10), GDK_MOD1_MASK,
            GTK_ACCEL_VISIBLE, goto_tab_closure[i]);
    }

    copy = g_cclosure_new_swap(G_CALLBACK(mainwindow_copy), 
        NULL, NULL);
    gtk_accel_group_connect(accel_group, 'c', conf_get_key_mod(),
        GTK_ACCEL_VISIBLE, copy);

    paste = g_cclosure_new_swap(G_CALLBACK(mainwindow_paste), 
        NULL, NULL);
    gtk_accel_group_connect(accel_group, 'v', conf_get_key_mod(),
        GTK_ACCEL_VISIBLE, paste);

    activetab = -1;
    tabcount = 0;
    GtkVBox* mainbox = GTK_VBOX(gtk_vbox_new(FALSE, 0));
    
    tabbar = GTK_NOTEBOOK(gtk_notebook_new());
    
    g_signal_connect(G_OBJECT(tabbar), "switch-page",
        G_CALLBACK(mainwindow_switch_tab), NULL);

    if(conf_get_opacity() < 100)
    {
        GdkScreen *screen = gdk_screen_get_default();
        GdkColormap *colormap = gdk_screen_get_rgba_colormap(screen);
        screen_is_composited = (colormap != NULL
            && gdk_screen_is_composited(screen));

        if(screen_is_composited)
        {
            gtk_widget_set_colormap(GTK_WIDGET(mainwindow), colormap);
            gdk_screen_set_default_colormap(screen, colormap);
        }
    }

    gtk_box_pack_start(GTK_BOX(mainbox), GTK_WIDGET(tabbar), TRUE, TRUE, 0);

    mainwindow_create_tab();

    gtk_widget_show_all(GTK_WIDGET(mainbox));
    gtk_container_add(GTK_CONTAINER(mainwindow), GTK_WIDGET(mainbox));

    int border = conf_get_border();
    if(border == BORDER_THIN)
        gtk_container_set_border_width(GTK_CONTAINER(mainwindow), 1);
    else if(border == BORDER_THICK)
        gtk_container_set_border_width(GTK_CONTAINER(mainwindow), 5);
    if(border != BORDER_NONE)
        g_signal_connect(G_OBJECT(mainwindow), "expose-event",
            G_CALLBACK(mainwindow_expose_event), NULL);

    if(conf_get_auto_hide())
        g_signal_connect(G_OBJECT(mainwindow), "focus-out-event",
            G_CALLBACK(mainwindow_focus_out_event), NULL);
    g_signal_connect(G_OBJECT(mainwindow), "show", G_CALLBACK(mainwindow_show), 
        NULL);
    g_signal_connect(G_OBJECT(mainwindow), "destroy",
        G_CALLBACK(mainwindow_destroy), NULL);

    gtk_signal_connect (GTK_OBJECT (mainwindow), "delete_event",
        GTK_SIGNAL_FUNC (delete_event), NULL);

    g_signal_connect_after(G_OBJECT(tabbar), "button_press_event", 
        G_CALLBACK(mainwindow_notebook_clicked), NULL);

    gtk_notebook_set_show_border(tabbar, FALSE);
    gtk_notebook_set_scrollable(tabbar, TRUE);
    if (conf_get_show_tab() == TABS_ONE|| conf_get_show_tab() == TABS_NEVER)
        gtk_notebook_set_show_tabs(tabbar, FALSE);
    gtk_notebook_set_tab_pos(tabbar, conf_get_tab_pos());
    gtk_notebook_set_homogeneous_tabs(tabbar, FALSE);

    XSetErrorHandler(handle_x_error);
    init_key();
    grab_key();
    g_thread_new("stjerm", (GThreadFunc)wait_key, NULL);
    
    // If stjerm has been started for the first time with --toggle, then
    // show the window straight away. Make haste!
    if(conf_get_toggled()) {
        mainwindow_toggle(1);
    }
}

void mainwindow_notebook_clicked(GtkWidget *widget, GdkEventButton *event, gpointer func_data)
{
    gint tabclicked = mainwindow_tab_at_xy(GTK_NOTEBOOK(widget), event->x, event->y);
    
    if(tabclicked > -1)
        gtk_notebook_set_current_page(GTK_NOTEBOOK(widget), tabclicked);
    else if(event->type == GDK_2BUTTON_PRESS)
        mainwindow_create_tab();
    
    if(event->button == 3)
    {
        popupmenu_shown = TRUE;
        gtk_menu_popup(GTK_MENU(popupmenu), NULL, NULL, NULL, NULL, event->button, event->time);
    }
}

/* This code adapted from gnome-terminal */
static gint mainwindow_tab_at_xy(GtkNotebook *notebook, gint x, gint y)
{
    GtkPositionType tab_pos;
    int page_num = 0;
    GtkWidget *page;
    
    tab_pos = gtk_notebook_get_tab_pos(notebook);
    
    if(notebook->first_tab == NULL)
        return -1;

    while((page = gtk_notebook_get_nth_page(notebook, page_num)))
    {
        GtkWidget *screen;
        gint max_x, max_y;

        screen = gtk_notebook_get_tab_label(notebook, page);
        g_return_val_if_fail(screen != NULL, -1);

        if(!GTK_WIDGET_MAPPED(GTK_WIDGET(screen)))
        {
            page_num++;
            continue;
        }

        max_x = screen->allocation.x + screen->allocation.width;
        max_y = screen->allocation.y + screen->allocation.height;

        if(((tab_pos == GTK_POS_TOP) || (tab_pos == GTK_POS_BOTTOM)) && (x <= max_x))
            return page_num;
        else if(((tab_pos == GTK_POS_LEFT) || (tab_pos == GTK_POS_RIGHT)) && (y <= max_y))
            return page_num;
        
        page_num++;
    }

    return -1;
}

void mainwindow_create_tab(void)
{
    GtkWidget* tmp_term = build_term();
    GtkVScrollbar *sbar= NULL;
    GtkHBox *tmp_box = GTK_HBOX(gtk_hbox_new(FALSE, 0));

    if(conf_get_scrollbar() == -1)
        gtk_box_pack_start(GTK_BOX(tmp_box), tmp_term, TRUE, TRUE, 0);
    else if(conf_get_scrollbar() == POS_LEFT)
    {
        sbar = GTK_VSCROLLBAR(gtk_vscrollbar_new(vte_terminal_get_adjustment(
            VTE_TERMINAL(tmp_term))));
        gtk_box_pack_start(GTK_BOX(tmp_box), GTK_WIDGET(sbar), FALSE, FALSE, 0);
        gtk_box_pack_end(GTK_BOX(tmp_box), GTK_WIDGET(tmp_term), TRUE, TRUE, 0);
    } 
    else // (conf_get_scrollbar() == POS_RIGHT)
    {
        sbar = GTK_VSCROLLBAR(gtk_vscrollbar_new(vte_terminal_get_adjustment(
            VTE_TERMINAL(tmp_term))));
        gtk_box_pack_start(GTK_BOX(tmp_box), GTK_WIDGET(tmp_term), TRUE, TRUE, 0);
        gtk_box_pack_end(GTK_BOX(tmp_box), GTK_WIDGET(sbar), FALSE, FALSE, 0);
    }

    char buffer [100];
    sprintf(buffer, "%s %d", conf_get_term_name(), activetab + 1);
    GtkLabel* tmp_label = GTK_LABEL(gtk_label_new(buffer));

    if(conf_get_opacity() < 100)
    {
        if(screen_is_composited)
        {
            vte_terminal_set_background_transparent(VTE_TERMINAL(tmp_term), FALSE);
            vte_terminal_set_opacity(VTE_TERMINAL(tmp_term),
                conf_get_opacity()/100 * 0xffff);
        }
        else
        {
            vte_terminal_set_background_saturation(VTE_TERMINAL(tmp_term),
                1.0 - conf_get_opacity()/100);
            if(conf_get_bg_image() == NULL)
                vte_terminal_set_background_transparent(VTE_TERMINAL(tmp_term), TRUE);
        }
    }

    if(conf_get_opacity() < 100 && screen_is_composited)
    {
        vte_terminal_set_background_transparent(VTE_TERMINAL(tmp_term), FALSE);
        vte_terminal_set_opacity(VTE_TERMINAL(tmp_term),
            conf_get_opacity()/100 * 0xffff);
    }
    
    g_signal_connect(G_OBJECT(tmp_term), "window-title-changed",
        G_CALLBACK(mainwindow_window_title_changed), tmp_label);

    tabcount++;

    gtk_widget_show_all(GTK_WIDGET(tmp_box));
    gtk_notebook_append_page(tabbar, GTK_WIDGET(tmp_box), GTK_WIDGET(tmp_label));

    if(conf_get_tab_fill())
        gtk_container_child_set(GTK_CONTAINER(tabbar), GTK_WIDGET(tmp_box),
            "tab-expand", TRUE, "tab-fill", TRUE, NULL);

    if(conf_get_show_tab() == TABS_ONE&& tabcount > 1)
        gtk_notebook_set_show_tabs(tabbar, TRUE);

    activetab = tabcount - 1;
    gtk_notebook_set_current_page(tabbar, activetab);

    if(conf_get_allow_reorder())
        gtk_notebook_set_tab_reorderable(tabbar, GTK_WIDGET(tmp_box), TRUE);
        
    guint i;
    
    for(i = 0; i < uri_regex_count; ++i)
    {
        int tag = vte_terminal_match_add_gregex(VTE_TERMINAL(tmp_term), uri_regex[i], 0);
        vte_terminal_match_set_cursor_type(VTE_TERMINAL(tmp_term), tag, GDK_HAND2);
    }
    
    if(tabcount > 1)
        gtk_widget_set_sensitive(close_tab, TRUE);
        
}

void mainwindow_close_tab(GtkWidget *term)
{
    // Look for the right tab...
    int thetab = activetab;
    
    if(term != NULL)
    {
        int i;
        
        // Loop each page in the notebook
        for (i = 0; i < gtk_notebook_get_n_pages(tabbar); i++) {
            // Grab the terminal from the notebook:
            GtkWidget *testTerm = mainwindow_get_terminal_at(i);

            if (term == testTerm) {
                thetab = i;
                break;
            }
        }
    }

    tabcount--;
    
    gtk_notebook_remove_page(tabbar, thetab);
    activetab = gtk_notebook_get_current_page(tabbar);

    if(tabcount == 1 && conf_get_show_tab() == TABS_ONE)
        gtk_notebook_set_show_tabs(tabbar, FALSE);
    
    if (tabcount == 0)
        mainwindow_create_tab();
    
    if(tabcount == 1)
        gtk_widget_set_sensitive(GTK_WIDGET(close_tab), FALSE);
}

void mainwindow_toggle(int sig)
{
    if((!sig && GTK_WIDGET_VISIBLE(mainwindow)) || (sig && toggled))
    {
        gdk_threads_enter();
        gtk_widget_hide(GTK_WIDGET(mainwindow));
        gdk_flush();
        gdk_threads_leave();
        toggled = FALSE;
        return;
    }
    
    toggled = TRUE;
    gdk_threads_enter();
    
    /* reset the window position before displaying it to avoid flickering if the
       windos is moved */
    mainwindow_reset_position();
    
    if(gtk_window_is_active(GTK_WINDOW(mainwindow)) == FALSE)
        gtk_window_present(GTK_WINDOW(mainwindow));
    else
        gtk_widget_show(mainwindow);
    
    gtk_window_stick(GTK_WINDOW(mainwindow));
    gtk_window_set_keep_above(GTK_WINDOW(mainwindow), TRUE);
    gdk_window_focus(mainwindow->window, gtk_get_current_event_time());
    gdk_flush();
    gdk_threads_leave();
}

void mainwindow_toggle_fullscreen(void)
{
    if(fullscreen)
    {
        gtk_window_unfullscreen(GTK_WINDOW(mainwindow));
        mainwindow_reset_position();
    }
    else
        gtk_window_fullscreen(GTK_WINDOW(mainwindow));
    
    fullscreen = !fullscreen;
    mainwindow_focus_terminal();
}

static void mainwindow_reset_position(void)
{
    int x, y;

    conf_get_position(&x, &y);
    gtk_window_move(GTK_WINDOW(mainwindow), x, y);
    gtk_window_resize(GTK_WINDOW(mainwindow), conf_get_width(), conf_get_height());
}

static void mainwindow_show(GtkWidget *widget, gpointer userdata)
{
    if(dpy != NULL)
        return;

    mw_xwin = GDK_WINDOW_XWINDOW(GTK_WIDGET(mainwindow)->window);
    dpy = GDK_DISPLAY_XDISPLAY(gdk_display_get_default());
}

static void mainwindow_focus_out_event(GtkWindow* window, GdkEventFocus* event, gpointer userdata)
{
    int revert;
    Window w;
    XGetInputFocus(dpy, &w, &revert);

    if (w == mw_xwin)
        return;

    toggled = FALSE;

    // focus wasn't lost just by pressing the shortcut key

    if(popupmenu_shown == TRUE)
        return;
        
    // focus wasn't lost by popping up popupmenu

    gtk_widget_hide(GTK_WIDGET(mainwindow));
}

static gboolean mainwindow_expose_event(GtkWidget *widget, GdkEventExpose *event, gpointer user_data)
{
    gint winw, winh;
    gtk_window_get_size(GTK_WINDOW(widget), &winw, &winh);

    gdk_draw_rectangle(widget->window, widget->style->black_gc, FALSE, 0, 0,
        winw-1, winh-1);

    if(conf_get_border() == BORDER_THIN)
        return FALSE;

    gdk_draw_rectangle(widget->window,
        widget->style->bg_gc[GTK_STATE_SELECTED], TRUE, 1, 1, winw -2, winh -2);

    gdk_draw_rectangle(widget->window, widget->style->bg_gc[GTK_STATE_NORMAL],
        TRUE, 5, 5, winw-10, winh-10);

    return FALSE;
}

static void mainwindow_destroy(GtkWidget *widget, gpointer user_data)
{
    gtk_main_quit();
}

static void mainwindow_window_title_changed(VteTerminal *vteterminal, gpointer user_data)
{
    if(vteterminal != NULL && user_data != NULL)
        gtk_label_set_label(GTK_LABEL(user_data), vte_terminal_get_window_title(vteterminal));
}

static void mainwindow_switch_tab(GtkNotebook *notebook, GtkNotebookPage *page,
    guint page_num, gpointer user_data)
{
    activetab = page_num;
}

static void mainwindow_goto_tab(gint i)
{
    gtk_notebook_set_current_page(tabbar, i);
    activetab = gtk_notebook_get_current_page(tabbar);
    mainwindow_focus_terminal();
}

static void mainwindow_next_tab(GtkWidget *widget, gpointer user_data)
{
    if(gtk_notebook_get_current_page(tabbar) == (tabcount - 1))
        gtk_notebook_set_current_page(tabbar, 0);
    else
        gtk_notebook_next_page(tabbar);

    activetab = gtk_notebook_get_current_page(tabbar);
    mainwindow_focus_terminal();
}

static void mainwindow_prev_tab(GtkWidget *widget, gpointer user_data)
{
    if(gtk_notebook_get_current_page(tabbar) == (0))
        gtk_notebook_set_current_page(tabbar, (tabcount - 1));
    else    
        gtk_notebook_prev_page(tabbar);

    activetab = gtk_notebook_get_current_page(tabbar);
    mainwindow_focus_terminal();
}

static void mainwindow_new_tab(GtkWidget *widget, gpointer user_data)
{
    mainwindow_create_tab();
    mainwindow_focus_terminal();
}

static void mainwindow_delete_tab(GtkWidget *widget, gpointer user_data)
{
    mainwindow_close_tab(NULL);

    if(tabcount > 0)
        mainwindow_focus_terminal();
}

int handle_x_error(Display *dpy, XErrorEvent *evt)
{
    if(evt->error_code == BadAccess|| evt->error_code == BadValue || 
        evt->error_code == BadWindow) 
    {
        fprintf(stderr, "error: unable to grab key, is stjerm is already running with the same key?\n");
        exit(1);
    }
    
    return 0;
}

static void mainwindow_focus_terminal(void)
{
    if (activetab >= 0)
        gtk_window_set_focus(GTK_WINDOW(mainwindow), mainwindow_get_terminal_at(activetab));
}

static gint mainwindow_copy(GtkWidget *widget, gpointer user_data)
{
    vte_terminal_copy_clipboard(
        VTE_TERMINAL(mainwindow_get_terminal_at(activetab))
    );
    return TRUE;
}

static gint mainwindow_paste(GtkWidget *widget, gpointer user_data) 
{
    vte_terminal_paste_clipboard(
        VTE_TERMINAL(mainwindow_get_terminal_at(activetab))
    );
    return TRUE;
}

GtkWidget* mainwindow_get_terminal_at(int index)
{
    GtkWidget *page = gtk_notebook_get_nth_page(tabbar, index);
    GtkWidget *term = NULL;

    GList *children = gtk_container_get_children(GTK_CONTAINER(page));
        
    if (conf_get_scrollbar() == POS_LEFT) {
        term = g_list_nth_data(children, 1);
    } else {
        term = g_list_nth_data(children, 0);
    }
    
    return term;
}

static gint delete_event(GtkWidget *widget, GdkEvent  *event, gpointer data)
{
    return(conf_ignore_delete_event());
}
